### PR TITLE
[5.2] Add getMultiple and putMultiple to cache implementations

### PR DESCRIPTION
**Pull request by nicholas-eden**
Primarily to save round-trips to redis and memcached when working with multiple objects

### CODE REVIEW

File: src/Illuminate/Cache/ApcStore.php [hunk 1]
```diff
@@ -49,6 +49,24 @@ public function get($key)
     }
 
     /**
+     * Retrieve multiple items from the cache by key,
```
Comment (GrahamCampbell) [inline src/Illuminate/Cache/ApcStore.php]: Please put a fullstop here, then a newline.

Comment (GrahamCampbell) [inline src/Illuminate/Cache/ApcStore.php]: So, like this:

```php
    /**
     * Retrieve multiple items from the cache by key.
     *
     * Items not found in the cache will have a null value for the key.
     *
     * @param  array  $keys
     * @return array
     */
```

Comment (GrahamCampbell) [inline src/Illuminate/Cache/ApcStore.php]: Also, note the changes I made to the param annotation in that snippet. Please do that too.

File: src/Illuminate/Cache/ApcStore.php [hunk 2]
```diff
@@ -62,6 +80,20 @@ public function put($key, $value, $minutes)
     }
 
     /**
+     * Store multiple items in the cache for a set number of minutes.
+     *
+     * @param array $values array of key => value pairs
```
Comment (GrahamCampbell) [inline src/Illuminate/Cache/ApcStore.php]: Please drop the comment here, and fix the alignment.

Comment (GrahamCampbell) [inline src/Illuminate/Cache/ApcStore.php]: So, like this:

```php
    /**
     * Store multiple items in the cache for a set number of minutes.
     *
     * @param  array  $values
     * @param  int  $minutes
     * @return void
     */
```

Comment (GrahamCampbell) [inline src/Illuminate/Cache/ApcStore.php]: Again, note the placement of spaces in the param annotations.

File: src/Illuminate/Cache/ArrayStore.php [hunk 3]
```diff
@@ -27,6 +27,24 @@ public function get($key)
     }
 
     /**
+     * Retrieve multiple items from the cache by key,
```
Comment (GrahamCampbell) [inline src/Illuminate/Cache/ArrayStore.php]: same comments apply for each occurence ;)

**Comment by nicholas-eden** [comment]
I can probably get around the phpunit version, 4.0 is missing the withConsecutive method.  When the build is done with lowest, it uses for 4.0, but stable uses 4.8.  Rather huge difference there.

File: src/Illuminate/Cache/NullStore.php [hunk 4]
```diff
@@ -25,6 +25,19 @@ public function get($key)
...
+     * Retrieve multiple items from the cache by key.
+     *
+     * Items not found in the cache will have a null value for the key.
+     *
+     * @param  array  $keys
+     * @return array
+     */
+    public function getMultiple(array $keys)
+    {
+        //
```
Comment (GrahamCampbell) [inline src/Illuminate/Cache/NullStore.php]: this needs to return an array

Comment (GrahamCampbell) [inline src/Illuminate/Cache/NullStore.php]: if keys is of length 3, then we should return `[null, null, null]`, right?

Comment (GrahamCampbell) [inline src/Illuminate/Cache/NullStore.php]: what you probably want here is:

```php
    public function getMultiple(array $keys)
    {
        $returnValues = [];

        foreach ($keys as $singleKey) {
            $returnValues[$singleKey] = $this->get($singleKey);
        }

        return $returnValues;
}
```

Comment (nicholas-eden) [inline src/Illuminate/Cache/NullStore.php]: That makes sense.  I wasn't sure how to handle the null store.  I'll update that.

Comment (GrahamCampbell) [inline src/Illuminate/Cache/NullStore.php]: or just `null` instead of `->get()`, but calling get is most correct

Comment (GrahamCampbell) [inline src/Illuminate/Cache/NullStore.php]: i guess you should really call put with put multiple too

Comment (nicholas-eden) [inline src/Illuminate/Cache/NullStore.php]: Updated that.  That method may have been more appropriate as a trait for the implementations that don't have a better way of handling getting/settings multiple values at once.

File: src/Illuminate/Cache/MemcachedStore.php [hunk 5]
```diff
@@ -49,6 +49,35 @@ public function get($key)
...
+    public function getMultiple(array $keys)
+    {
+        $prefixedKeys = [];
+
+        foreach ($keys as $keyToPrefix) {
+            $prefixedKeys[] = $this->prefix.$keyToPrefix;
+        }
+
+        $cas = null;
+        $flags = 1; //\Memcached::GET_PRESERVE_ORDER
```
Comment (GrahamCampbell) [inline src/Illuminate/Cache/MemcachedStore.php]: Please format this comment as just `// Memcached::GET_PRESERVE_ORDER`

Comment (GrahamCampbell) [inline src/Illuminate/Cache/MemcachedStore.php]: wait, what?

Comment (GrahamCampbell) [inline src/Illuminate/Cache/MemcachedStore.php]: Why arn't you use the actual constant?

File: tests/Pipeline/PipelineTest.php [hunk 6]
```diff
@@ -29,7 +29,7 @@ public function testPipelineBasicUsage()
 
     public function testPipelineUsageWithParameters()
     {
-        $parameters = ['one','two'];
+        $parameters = ['one', 'two'];
```
Comment (GrahamCampbell) [inline tests/Pipeline/PipelineTest.php]: Please rebase to get rid of this change.

File: src/Illuminate/Cache/DatabaseStore.php [hunk 7]
```diff
@@ -86,6 +86,25 @@ public function get($key)
...
 
     /**
+     * Retrieve multiple items from the cache by key.
+     *
+     * Items not found in the cache will have a null value for the key.
+     *
+     * @param  array  $keys
+     * @return array
+     */
+    public function getMultiple(array $keys)
```
Comment (GrahamCampbell) [inline src/Illuminate/Cache/DatabaseStore.php]: This stuff is common to multiple stores. Maybe we need a trait here?

File: src/Illuminate/Cache/MemcachedStore.php (hunk 5 above)
Comment (nicholas-eden) [inline src/Illuminate/Cache/MemcachedStore.php]: I did initially, it failed the test due to the module not being installed.  How do you suggest handling this?

File: src/Illuminate/Cache/DatabaseStore.php (hunk 7 above)
Comment (nicholas-eden) [inline src/Illuminate/Cache/DatabaseStore.php]: I was thinking that as well.  Should the trait go in that same folder?  That seems to be the convention.

File: src/Illuminate/Cache/MemcachedStore.php (hunk 5 above)
Comment (GrahamCampbell) [inline src/Illuminate/Cache/MemcachedStore.php]: Skip that test if the extension is not installed.

File: src/Illuminate/Cache/DatabaseStore.php (hunk 7 above)
Comment (GrahamCampbell) [inline src/Illuminate/Cache/DatabaseStore.php]: Yeh. If not, Taylor will ask it to be moved later, or just move it himself.

**Comment by nicholas-eden** [comment]
It is still failing me for that Pipeline file for some reason.

**Comment by GrahamCampbell** [comment]
> It is still failing me for that Pipeline file for some reason.

Please rebase and it won't fail anymore. I fixed that earlier today

**Comment by nicholas-eden** [comment]
Ah, I was rebasing to my fork rather than laravel.  Still getting the hang of working with forks.

File: composer.json [hunk 8]
```diff
@@ -76,7 +76,7 @@
         "iron-io/iron_mq": "~2.0",
         "mockery/mockery": "~0.9.2",
         "pda/pheanstalk": "~3.0",
-        "phpunit/phpunit": "~4.0",
+        "phpunit/phpunit": "~4.1",
```
Comment (ArjanSchouten) [inline composer.json]: Why are you updating phpunit within this PR? If phpunit will be updated let's move this to another PR?!

Comment (nicholas-eden) [inline composer.json]: Sure, I'll submit one later. The conversation on that got collapsed from the rebase. There are method missing in 4.0 that a test needs.

Comment (ArjanSchouten) [inline composer.json]: Alright :) missed that!

**Comment by taylorotwell** [comment]
Merged this but did quite a bit of refactoring. You can just do:

```php
Cache::get(['foo', 'bar' => 'default-value']);

Cache::put(['foo' => 'value'], 10);
```

**Comment by GrahamCampbell** [comment]
Sexy. :)
